src/backend/TxCache.cpp

```diff
@@ -0,0 +1,105 @@
+#include <backend/TxCache.h>
+namespace Backend {
+
+uint32_t
+TxCache::latestLedgerSequence() const
+{
+    std::shared_lock lck{mtx_};
+    return latestSeq_;
+}
+
+// adds an entire ledger's worth of transactions to the cache. Evicts oldest
+// ledger.
+void
+TxCache::update(
+    std::vector<ripple::uint256> const& hashes,
+    std::vector<TransactionAndMetadata> const& transactions,
+    uint32_t seq)
+{
+    std::unique_lock lck{mtx_};
+    cache_[tail_].clear();
+
+    assert(seq == latestSeq_ + 1 || latestSeq_ == 0);
+    latestSeq_ = seq;
+    for (std::size_t i = 0; i < hashes.size(); i++)
+    {
+        cache_[tail_][hashes[i]] = transactions[i];
+    }
+    tail_ = (tail_ + 1) % NUM_LEDGERS_CACHED;
+}
+
+std::optional<TransactionAndMetadata>
+TxCache::get(ripple::uint256 const& key) const
+{
+    std::shared_lock lck{mtx_};
+    txReqCounter_++;
+
+    for (auto& map : cache_)
+    {
+        auto e = map.find(key);
+        if (e != map.end())
+        {
+            txHitCounter_++;
+            return {e->second};
+        }
+    }
+    return {};
```

Comment:
so if we had a second map, that mapped a transaction hash to a ledger sequence, we wouldn't need to do this loop. The loop is probably not so bad for a small number of ledgers, but for a larger number of ledgers, maybe a second map would make sense. What do you think?